### PR TITLE
bind dhcp sockets to IP addresses instead of the interfaces

### PR DESCRIPTION
**Pull request by yegor-alexeyev**
Proposed fix for https://github.com/systemd/systemd/issues/18939

### CODE REVIEW

**Comment by poettering** [comment]
I am not sure how I feel about this. This looks like a security issue, if a DHCP server that is bound to one interface suddenly responds on another. I don't grok the usecase I must say.

**Comment by poettering** [comment]
I guess we can make this configurable in some way, but I'd ensure that by default we clearly isolate our networks. This is per-network configuration after all. The setup mentioned in the original issue appears to be the outlier really, and not how DHCP is typically used.

**Comment by yegor-alexeyev** [comment]
Hi! I updated the pull request. BindToInterface=bool DHCP server configuration option is added. It regulates how to bind the server socket. It's "true" by default, so by default all works the old way.
The default codeflow is almost unaffected(the only change I see is that the dhcp client socket now getting SO_BROADCAST option. I see nothing bad about it).

The implementation is simple, When BindToInterface set to "no", interface bindings and IP_PKTINFO filters are not used. Since socket bound to usual IP-address is not able to receive broadcasts, another interface bound broadcast socket is used to receive broadcasts.
 
I tested the changes, all seems good and works as expected. Particularly the sockets are still local to interfaces, and there is no problem with launching multiple dhcp-server instances each on its own interface.

File: src/libsystemd-network/dhcp-network.c [hunk 1]
```diff
@@ -186,19 +186,20 @@ int dhcp_network_bind_udp_socket(int ifindex, be32_t address, uint16_t port, int
                         return r;
         }
 
+        r = setsockopt_int(s, SOL_SOCKET, SO_BROADCAST, true);
+        if (r < 0)
+                return r;
```
Comment (poettering) [inline src/libsystemd-network/dhcp-network.c]: can you explain why you moved this?

File: src/libsystemd-network/dhcp-network.c [hunk 2]
```diff
@@ -186,19 +186,20 @@ int dhcp_network_bind_udp_socket(int ifindex, be32_t address, uint16_t port, int
...
-
-                r = setsockopt_int(s, SOL_SOCKET, SO_BROADCAST, true);
-                if (r < 0)
-                        return r;
-
         } else {
-                r = setsockopt_int(s, IPPROTO_IP, IP_FREEBIND, true);
-                if (r < 0)
-                        return r;
+                if (address != INADDR_BROADCAST) {
```
Comment (poettering) [inline src/libsystemd-network/dhcp-network.c]: in order to keep indentation levels low, please merge with the `else` to become `} else if (address…`

File: src/libsystemd-network/dhcp-network.c [hunk 3]
```diff
@@ -186,19 +186,20 @@ int dhcp_network_bind_udp_socket(int ifindex, be32_t address, uint16_t port, int
                         return r;
         }
 
+        r = setsockopt_int(s, SOL_SOCKET, SO_BROADCAST, true);
+        if (r < 0)
+                return r;
+
         if (address == INADDR_ANY) {
                 r = setsockopt_int(s, IPPROTO_IP, IP_PKTINFO, true);
```
Comment (poettering) [inline src/libsystemd-network/dhcp-network.c]: please add comment here, why this is not necessary in your new mode

File: src/libsystemd-network/sd-dhcp-server.c [hunk 4]
```diff
@@ -232,9 +233,17 @@ int sd_dhcp_server_stop(sd_dhcp_server *server) {
         server->receive_message =
                 sd_event_source_unref(server->receive_message);
 
+        if (!server->bind_to_interface) {
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-server.c]: please drop this if check. we want destructors to be as robust as possible, hence if there's an object set here, destroy it. All our xyz_unref() functions become NOPs if you pass NULL to them, to make this easy and safe.

File: src/libsystemd-network/sd-dhcp-server.c [hunk 5]
```diff
@@ -232,9 +233,17 @@ int sd_dhcp_server_stop(sd_dhcp_server *server) {
...
+                server->receive_broadcast =
+                        sd_event_source_unref(server->receive_broadcast);
+        }
+
         server->fd_raw = safe_close(server->fd_raw);
         server->fd = safe_close(server->fd);
 
+        if (!server->bind_to_interface)
+                server->fd_broadcast = safe_close(server->fd_broadcast);
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-server.c]: as above, drop the if check, destroy the fd no matter what. safe_close() is a NOP if you pass < 0, to make this safe and easy

File: src/libsystemd-network/sd-dhcp-server.c [hunk 6]
```diff
@@ -282,9 +291,10 @@ static int dhcp_server_send_udp(sd_dhcp_server *server, be32_t destination,
                 .msg_namelen = sizeof(dest.in),
                 .msg_iov = &iov,
                 .msg_iovlen = 1,
-                .msg_control = &control,
-                .msg_controllen = sizeof(control),
+                .msg_control = NULL,
+                .msg_controllen = 0,
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-server.c]: structured initialization in C defaults to zero initialize unlisted fields, hence just drop these two lines

File: src/libsystemd-network/sd-dhcp-server.c [hunk 7]
```diff
@@ -997,7 +1012,12 @@ int sd_dhcp_server_start(sd_dhcp_server *server) {
         }
         server->fd_raw = r;
 
-        r = dhcp_network_bind_udp_socket(server->ifindex, INADDR_ANY, DHCP_PORT_SERVER, -1);
+        if (server->bind_to_interface) {
+                r = dhcp_network_bind_udp_socket(server->ifindex, INADDR_ANY, DHCP_PORT_SERVER, -1);
+        } else {
+                r = dhcp_network_bind_udp_socket(0, server->address, DHCP_PORT_SERVER, -1);
+        }
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-server.c]: as per coding style drop {} around single-line if blocks:

https://systemd.io/CODING_STYLE/#formatting

File: src/libsystemd-network/sd-dhcp-server.c [hunk 8]
```diff
@@ -1019,6 +1039,33 @@ int sd_dhcp_server_start(sd_dhcp_server *server) {
                 return r;
         }
 
+
+        if (!server->bind_to_interface) {
+                r = dhcp_network_bind_udp_socket(server->ifindex, INADDR_BROADCAST, DHCP_PORT_SERVER, -1);
+
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-server.c:1049]: please drop this empty line, we generally try to keep function calls and their immediate error handling together, which follows here, hence there should not be an empty line in-between

File: src/libsystemd-network/sd-dhcp-server.c [hunk 9]
```diff
@@ -1019,6 +1039,33 @@ int sd_dhcp_server_start(sd_dhcp_server *server) {
...
+                if (r < 0) {
+                        sd_dhcp_server_stop(server);
+                        return r;
+                }
+
+                r = sd_event_source_set_priority(server->receive_broadcast,
+                                                 server->event_priority);
+                if (r < 0) {
+                        sd_dhcp_server_stop(server);
+                        return r;
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-server.c]: one of those days we should replace this error path by a goto label at the end of this function that calls `sd_dhcp_server_stop()` and returns r, we do this so often already in this function

File: src/libsystemd-network/sd-dhcp-server.c [hunk 10]
```diff
@@ -1019,6 +1039,33 @@ int sd_dhcp_server_start(sd_dhcp_server *server) {
...
+                }
+        }
+
+
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-server.c]: spurious double newline

File: src/libsystemd-network/sd-dhcp-server.c [hunk 11]
```diff
@@ -1048,6 +1095,17 @@ int sd_dhcp_server_forcerenew(sd_dhcp_server *server) {
         return r;
 }
 
+int sd_dhcp_server_set_bind_to_interface(sd_dhcp_server *server, int enabled) {
+        assert_return(server, -EINVAL);
+
+        if (enabled == server->bind_to_interface)
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-server.c]: so, this is supposed to be a public API, and in public APIs we so far exposed booleans as "int" for compt with pre-C99. internally we use "bool" everywhere however. This is mostly OK everywhere and the distinction doesn't matter much. But when comparing the two types we need to take special care since passing "2" in such a bool than is stored as "int" should be treated as true, but here it is treated as something that compares neither with true nor with false.

Hence, please, when comparing the parameter here, use `!!enabled == server->bind_to_interface`.

File: man/systemd.network.xml [hunk 12]
```diff
@@ -2411,6 +2411,14 @@ IPv6Token=prefixstable:2002:da8:1::</programlisting></para>
           then all options specified earlier are cleared. Defaults to unset.</para>
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>BindToInterface=</varname></term>
+        <listitem>Takes a boolean value. When <literal>yes</literal>, DHCP server socket will be bound
+        to its network interface and all socket communication will be restricted to this interface.
+        Defaults to <literal>yes</literal>.
+        Disabling interface binding is mostly useful in the networks with DHCP relays.
```
Comment (poettering) [inline man/systemd.network.xml]: this last sentence confuses me more than clear things up. i mean, it only applies to this setup of yours where you use distinct interfaces for this relaying. maybe reword?

**Review by poettering — CHANGES_REQUESTED** [review]
i leave the final review to @yuwata

File: man/systemd.network.xml (hunk 12 above)
Comment (yegor-alexeyev) [reply man/systemd.network.xml]: Okay, I better remove that sentence at all.

File: src/libsystemd-network/dhcp-network.c (hunk 1 above)
Comment (yegor-alexeyev) [reply src/libsystemd-network/dhcp-network.c]: Sending broadcast is required for dhcp-server socket, regardless of how it's bound.

Comment (yegor-alexeyev) [reply src/libsystemd-network/dhcp-network.c]: Moving this outside of the condition makes dhcp-client socket to have SO_BROADCAST too(unlike current implementation). But I thought that adding check like `if (port == DHCP_PORT_SERVER)`  or additional function parameter will be more confusing then useful.

**Comment by yegor-alexeyev** [comment]
> i leave the final review to @yuwata

Hi Lennart. Thanks for the review, I updated the pull request.

File: src/libsystemd-network/sd-dhcp-server.c [hunk 13]
```diff
@@ -997,7 +1007,11 @@ int sd_dhcp_server_start(sd_dhcp_server *server) {
         }
         server->fd_raw = r;
 
-        r = dhcp_network_bind_udp_socket(server->ifindex, INADDR_ANY, DHCP_PORT_SERVER, -1);
+        if (server->bind_to_interface)
+                r = dhcp_network_bind_udp_socket(server->ifindex, INADDR_ANY, DHCP_PORT_SERVER, -1);
+        else
+                r = dhcp_network_bind_udp_socket(0, server->address, DHCP_PORT_SERVER, -1);
+
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-server.c]: we generally try to keep function calls and their immediate error handling together, hence please remove the empty line here that separates these functions from their error handling

File: src/libsystemd-network/sd-dhcp-server.c [hunk 14]
```diff
@@ -176,6 +176,7 @@ int sd_dhcp_server_new(sd_dhcp_server **ret, int ifindex) {
         server->n_ref = 1;
         server->fd_raw = -1;
         server->fd = -1;
+        server->fd_broadcast = -1;
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-server.c]: bind_to_interface should defualt to true here, too

Comment (yegor-alexeyev) [reply src/libsystemd-network/sd-dhcp-server.c]: Thanks, I missed this one. All done

File: src/libsystemd-network/sd-dhcp-server.c [hunk 15]
```diff
@@ -1048,6 +1086,17 @@ int sd_dhcp_server_forcerenew(sd_dhcp_server *server) {
         return r;
 }
 
+int sd_dhcp_server_set_bind_to_interface(sd_dhcp_server *server, int enabled) {
+        assert_return(server, -EINVAL);
```
Comment (yuwata) [inline src/libsystemd-network/sd-dhcp-server.c:1098]: Maybe, we should refuse this when the server is already started.

File: src/libsystemd-network/sd-dhcp-server.c (hunk 9 above)
Comment (yuwata) [reply src/libsystemd-network/sd-dhcp-server.c]: Not addressed yet.

**Review by yuwata — CHANGES_REQUESTED** [review]
Not fully reviewed yet. Only cosmetic comments.

File: src/libsystemd-network/sd-dhcp-server.c (hunk 15 above)
Comment (yegor-alexeyev) [reply src/libsystemd-network/sd-dhcp-server.c:1098]: Hi! Right, the way as is it will just break the server. I added a check for this.

**Comment by yegor-alexeyev** [comment]
> Not fully reviewed yet. Only cosmetic comments.
Hello! Any updates about this?

File: src/libsystemd-network/dhcp-network.c [hunk 16]
```diff
@@ -186,16 +186,18 @@ int dhcp_network_bind_udp_socket(int ifindex, be32_t address, uint16_t port, int
                         return r;
         }
 
+        r = setsockopt_int(s, SOL_SOCKET, SO_BROADCAST, true);
```
Comment (yuwata) [inline src/libsystemd-network/dhcp-network.c]: Please add the condition `if (port == DHCP_PORT_SERVER)`.

File: src/libsystemd-network/dhcp-network.c [hunk 17]
```diff
@@ -186,16 +186,18 @@ int dhcp_network_bind_udp_socket(int ifindex, be32_t address, uint16_t port, int
...
                 if (r < 0)
                         return r;
-
-                r = setsockopt_int(s, SOL_SOCKET, SO_BROADCAST, true);
-                if (r < 0)
-                        return r;
-
-        } else {
+        } else if (address != INADDR_BROADCAST) {
                 r = setsockopt_int(s, IPPROTO_IP, IP_FREEBIND, true);
```
Comment (yuwata) [inline src/libsystemd-network/dhcp-network.c:202]: Originally, this option is for dhcp client. Please change the condition to `if (port != DHCP_PORT_SERVER)`.

File: src/libsystemd-network/sd-dhcp-server.c [hunk 18]
```diff
@@ -253,9 +255,14 @@ int sd_dhcp_server_stop(sd_dhcp_server *server) {
         server->receive_message =
                 sd_event_source_unref(server->receive_message);
 
+        server->receive_broadcast =
+                sd_event_source_unref(server->receive_broadcast);
```
Comment (yuwata) [inline src/libsystemd-network/sd-dhcp-server.c]: Not necessary to wrap such short lines.

File: src/libsystemd-network/sd-dhcp-server.c [hunk 19]
```diff
@@ -253,9 +255,14 @@ int sd_dhcp_server_stop(sd_dhcp_server *server) {
...
+
         server->fd_raw = safe_close(server->fd_raw);
         server->fd = safe_close(server->fd);
 
+        server->fd_broadcast = safe_close(server->fd_broadcast);
```
Comment (yuwata) [inline src/libsystemd-network/sd-dhcp-server.c]: Not necessary to add a blank line.

File: src/libsystemd-network/sd-dhcp-server.c [hunk 20]
```diff
@@ -1018,7 +1028,10 @@ int sd_dhcp_server_start(sd_dhcp_server *server) {
         }
         server->fd_raw = r;
 
-        r = dhcp_network_bind_udp_socket(server->ifindex, INADDR_ANY, DHCP_PORT_SERVER, -1);
+        if (server->bind_to_interface)
+                r = dhcp_network_bind_udp_socket(server->ifindex, INADDR_ANY, DHCP_PORT_SERVER, -1);
+        else
+                r = dhcp_network_bind_udp_socket(0, server->address, DHCP_PORT_SERVER, -1);
         if (r < 0) {
                 sd_dhcp_server_stop(server);
```
Comment (yuwata) [inline src/libsystemd-network/sd-dhcp-server.c]: Please introduce a goto label to stop server on error, e.g.,
```
r = socket(...);
if (r < 0) {
    r = -errno;
    goto on_error;
}

if (server->bind_to_interface)
    r = dhcp_network_bind_udp_socket(...);
else
    r = dhcp_network_bind_udp_socket(...);
if (r < 0)
    goto on_error;

...

on_error:
    sd_dhcp_server_stop();
    return r;
```

File: src/libsystemd-network/sd-dhcp-server.c [hunk 21]
```diff
@@ -1069,6 +1107,22 @@ int sd_dhcp_server_forcerenew(sd_dhcp_server *server) {
...
+int sd_dhcp_server_set_bind_to_interface(sd_dhcp_server *server, int enabled) {
+        assert_return(server, -EINVAL);
+
+        if (!!enabled == server->bind_to_interface)
+                return 0;
+
+        if (sd_dhcp_server_is_running(server)) {
+                log_dhcp_server(server, "Refusing to change BindToInterface setting on a running DHCP Server. Restart is needed");
+                return 0;
+        }
```
Comment (yuwata) [inline src/libsystemd-network/sd-dhcp-server.c]: I think we should refuse and return error in such case. ie
```
assert_return(!sd_dhcp_server_is_running(server), -EBUSY);
```

File: src/network/networkd-network.c [hunk 22]
```diff
@@ -382,6 +382,8 @@ int network_load_one(Manager *manager, OrderedHashmap **networks, const char *fi
                 .dhcp6_pd_manage_temporary_address = true,
                 .dhcp6_pd_subnet_id = -1,
 
+                .dhcp_server_bind_to_interface = true,
+
```
Comment (yuwata) [inline src/network/networkd-network.c]: Please drop the blank line. Could you also drop the blank line above the line `.dhcp_server_emit_router = true;`?

**Review by yuwata — CHANGES_REQUESTED** [review]
Sorry for late review.

File: src/libsystemd-network/dhcp-network.c (hunk 16 above)
Comment (yegor-alexeyev) [reply src/libsystemd-network/dhcp-network.c]: Sure, I thought about this too

**Comment by yegor-alexeyev** [comment]
> Sorry for late review.

It's ok. I made the edits you requested.

File: src/libsystemd-network/dhcp-network.c [hunk 23]
```diff
@@ -186,16 +186,20 @@ int dhcp_network_bind_udp_socket(int ifindex, be32_t address, uint16_t port, int
...
-        if (address == INADDR_ANY) {
-                r = setsockopt_int(s, IPPROTO_IP, IP_PKTINFO, true);
+        if (port == DHCP_PORT_SERVER) {
+                r = setsockopt_int(s, SOL_SOCKET, SO_BROADCAST, true);
                 if (r < 0)
                         return r;
+        }
 
-                r = setsockopt_int(s, SOL_SOCKET, SO_BROADCAST, true);
+        if (address == INADDR_ANY) {
```
Comment (yuwata) [inline src/libsystemd-network/dhcp-network.c]: Hm, the conditions become complicated now.
How about the following:
```
if (port == DHCP_PORT_SERVER) {
        r = setsockopt_int(s, SOL_SOCKET, SO_BROADCAST, true);
        if (r < 0)
                return r;
        if (address == INADDR_ANY) {
                r = setsockopt_int(s, IPPROTO_IP, IP_PKTINFO, true);
                if (r < 0)
                        return r;
        }
} else {
        r = setsockopt_int(s, IPPROTO_IP, IP_FREEBIND, true);
        if (r < 0)
                return r;
}
```

File: src/libsystemd-network/sd-dhcp-server.c [hunk 24]
```diff
@@ -1013,36 +1019,56 @@ int sd_dhcp_server_start(sd_dhcp_server *server) {
...
 
         r = sd_event_source_set_priority(server->receive_message,
                                          server->event_priority);
-        if (r < 0) {
-                sd_dhcp_server_stop(server);
-                return r;
+        if (r < 0)
+                goto on_error;
+
+
```
Comment (yuwata) [inline src/libsystemd-network/sd-dhcp-server.c]: Doubled empty lines.

**Review by yuwata — CHANGES_REQUESTED** [review]
Looks mostly good. Only cosmetic issues.

File: src/libsystemd-network/dhcp-network.c (hunk 23 above)
Comment (yegor-alexeyev) [reply src/libsystemd-network/dhcp-network.c]: Right, client and server codepath is distinguished now. Done

**Comment by yuwata** [comment]
Could you squash commits and drop merge commit in this PR? Some of our CIs cannot handle merge commits in PR...

**Comment by yegor-alexeyev** [comment]
> LGTM.
> 
> Could you squash commits and drop merge commit in this PR? Some of our CIs cannot handle merge commits in PR...

Done, rebased and squashed the changes to a single commit.
